src/element-service.js

```diff
@@ -185,31 +186,46 @@ function assertService(service, id, extension) {
 }
 
 /**
- * Waits for an extension if a stub is present
+ * Get list of all the extension JS files
+ * @param {HTMLHeadElement|Element|ShadowRoot} head
+ * @return {!Array<string>}
+ */
+export function extensionScriptsInNode(head) {
+  // ampdoc.getHeadNode() can return null
+  if (!head) {
+    return [];
+  }
+  const scripts = [];
+  const list = head.querySelectorAll('script[custom-element]');
+  for (let i = 0; i < list.length; i++) {
+    scripts.push(list[i].getAttribute('custom-element'));
+  }
+  return scripts;
+}
+
+/**
+ * Waits for an extension if its script is present
  * @param {!Window} win
  * @param {string} extension
+ * @param {!Array<string>} customElements
  * @return {!Promise}
  * @private
  */
-function waitForExtensionIfStubbed(win, extension) {
+function waitForExtensionIfPresent(win, extension, customElements) {
```

Comment:
At this point, this is just a more expensive `isElementScheduled`. There are very subtle areas where `isElementScheduled` can fail to detect an element, but it's [very minimal](https://github.com/ampproject/amphtml/blob/01a375256f635051a06a3b8113ac9ae63a37e8cb/src/amp.js#L76-L90).

It'd much rather see [`stubElementsForDoc`](https://github.com/ampproject/amphtml/blob/acc1927661eec3e7bea03456971c2cb7f83ffe4d/src/service/custom-element-registry.js#L97-L104) be split into two phases. The first will be a mark phase, adding everything to `ampExtendedElements`. The second will loop through `ampExtendedElements` to [add things to the ampdoc](https://github.com/ampproject/amphtml/blob/acc1927661eec3e7bea03456971c2cb7f83ffe4d/src/service/custom-element-registry.js#L101) when it's available. That eliminates any race, and removes the duplication here.

Reply:
It's the very subtle cases I'm concerned about.  At tis point my questions do we want to ship this given that we know it will work reliably, if not as efficiently as possible or are we prepared to accept an unknown failure rate.  (unknown because we don't log `user().error()`).

My take is we should make it work reliably then make it work efficiently.

Reply:
My approach above will eliminate the edge cases.

Reply:
@jridgewell  I'm not familiar enough with the flow here, particularly with shadow amp.  This has gone way beyond the scope of what I was trying to fix.  Given that I have another project to to deal with can I punt this to you?

Reply:
It's unclear. I saw the original race often enough when developing amp-geo to be concerned.  Adding the extra micro task has either fixed it or made it very rare on chrome.   The problem is when it fails in the real world it fails as a user().assert() and we don't see a log so I have no idea how often it's happening, if at all.